searchlib/src/vespa/searchlib/features/closest_feature.cpp

```diff
@@ -0,0 +1,288 @@
+};
+
+class ClosestSerializedExecutor : public ClosestExecutor {
+public:
+    ClosestSerializedExecutor(DistanceCalculatorBundle&& bundle, Value& empty_output, TypedCells identity, const ITensorAttribute& attr);
+    ~ClosestSerializedExecutor() override;
+    void execute(uint32_t docId) override;
+};
+
+class ClosestDirectExecutor : public ClosestExecutor {
+    SubspaceType            _subspace_type;
+    std::vector<string_id>  _labels;
+    std::vector<string_id*> _label_ptrs;
+public:
+    ClosestDirectExecutor(DistanceCalculatorBundle&& bundle, Value& empty_output, TypedCells identity, const ITensorAttribute& attr);
+    ~ClosestDirectExecutor() override;
+    void execute(uint32_t docId) override;
+};
+
+ClosestExecutor::ClosestExecutor(DistanceCalculatorBundle&& bundle, Value& empty_output, TypedCells identity, const ITensorAttribute& attr)
+    : _bundle(std::move(bundle)),
+      _empty_output(empty_output),
+      _identity(identity),
+      _attr(attr),
+      _output()
+{
+}
+
+ClosestExecutor::~ClosestExecutor() = default;
+
+fef::FeatureExecutor&
+ClosestExecutor::make(DistanceCalculatorBundle&& bundle, Value& empty_output, TypedCells identity, const ITensorAttribute& attr, vespalib::Stash& stash)
+{
+    if (attr.supports_get_serialized_tensor_ref()) {
+        return stash.create<ClosestSerializedExecutor>(std::move(bundle), empty_output, identity, attr);
+    } else if (attr.supports_get_tensor_ref()) {
+        return stash.create<ClosestDirectExecutor>(std::move(bundle), empty_output, identity, attr);
+    } else {
+        return ConstantTensorExecutor::createEmpty(empty_output.type(), stash);
+    }
+}
+
+ClosestSerializedExecutor::ClosestSerializedExecutor(DistanceCalculatorBundle&& bundle, Value& empty_output, TypedCells identity, const ITensorAttribute& attr)
+    : ClosestExecutor(std::move(bundle), empty_output, identity, attr)
+{
+}
+
+ClosestSerializedExecutor::~ClosestSerializedExecutor() = default;
+
+void
+ClosestSerializedExecutor::execute(uint32_t docId)
+{
+    double best_distance = 0.0;
+    std::optional<uint32_t> closest_subspace;
+    auto ref = _attr.get_serialized_tensor_ref(docId);
+    for (const auto& elem : _bundle.elements()) {
+        elem.calc->calc_closest_subspace(ref.get_vectors(), closest_subspace, best_distance);
+    }
+    if (closest_subspace.has_value()) {
+        auto labels = ref.get_labels(closest_subspace.value());
+        _output = std::make_unique<FastValueView>(_empty_output.type(), labels, _identity, labels.size(), 1);
+        outputs().set_object(0, *_output);
+    } else {
+        outputs().set_object(0, _empty_output);
+    }
+}
+
+ClosestDirectExecutor::ClosestDirectExecutor(DistanceCalculatorBundle&& bundle, Value& empty_output, TypedCells identity, const ITensorAttribute& attr)
+    : ClosestExecutor(std::move(bundle), empty_output, identity, attr),
+      _subspace_type(attr.getTensorType()),
+      _labels(attr.getTensorType().count_mapped_dimensions()),
```

Comment:
This can be simplified with stricter validation in setup(). It will always be 1.